examples/matrix_multiplication.masm

```diff
@@ -0,0 +1,727 @@
+#   [0, 0, 6, 9, 0, 0, 0, 0]
+#   [0, 0, 7, 1, 0, 0, 0, 0]
+#
+# Resulting in the following encoding:
+#
+# [
+#     0, 0, 3, 2, 0, 0, 0, 0,  # matrix size
+#     0, 0, 3, 8, 0, 0, 0, 0,  # first column
+#     0, 0, 6, 9, 0, 0, 0, 0,  # second column
+#     0, 0, 7, 1, 0, 0, 0, 0   # third column
+# ]
+#
+# The data is read into memory preserving padding *elements*, but removing
+# padding *words*. So the data above will be store as:
+#
+#     addr+0: [2, 3, 0, 0]  # matrix size
+#     addr+1: [8, 3, 0, 0]  # first column
+#     addr+2: [9, 6, 0, 0]  # second column
+#     addr+3: [1, 7, 0, 0]  # third column
+#
+# Notes:
+# - This procedure *must* be called with `exec` to copy the data to memory.
+# - The data packingig done to take advantage of `adv_pipe`, a very efficient
+# primitive to load and verify data from tape to memory. Tightly packing the
+# data into the tape would likely be two orders of magnitude slower.
+#
+proc.matrix_column_major_copy.0
+    # input: [addr, ...]
+    # output: [H, H, H, ...]
+
+    # [H, H, H, addr, 0, 0, 0, ...]
+    push.0 push.0 push.0 movup.3 padw padw padw
+
+    adv_pipe
+
+    # [addr, 0, 0, 0, H, H, H, ...]
+    movupw.3
+
+    # remove the padding word after the [row, column]
+    sub.1
+
+    # [addr, 0, 0, 0, addr, H, H, H, ...]
+    dup movdn.4
+
+    # [row, col, 0, 0, addr, H, H, H, ...]
+    dup sub.1 mem_loadw
+
+    # [row, col, addr, H, H, H, ...]
+    movup.2 drop movup.2 drop
+
+    # Outter loop. The value `col` won't be needed again, so reuse it as a loop counter `c`.
+    # OUTER: [c, words, addr, H, H, H, ...]
+    exec.count_to_word swap
+
+    # if(c>0)
+    dup neq.0 while.true
+
+        # save a copy of `words` for the next iteration, call the loop counter `r`
+        # [r, words, c, addr, H, H, H, ...]
+        swap dup
+
+        # INNER: [r, addr, words, c, H, H, H, ...]
+        movup.3 swap
+
+        # if(r>0)
+        dup neq.0 while.true
+            # [H, H, H, addr, r, words, c, ...]
+            swap movdnw.3
+
+            # [H, H, H, addr, r, words, c, ...]
+            adv_pipe
+
+            # INNER: [r, addr, words, c, H, H, H, ...]
+            movupw.3 swap
+
+            # every call to `adv_pipe` copies *two* words
+            sub.2
+
+            # while(c>0)
+            dup exec.is_not_zero_or_negative_one
+        end
+
+        # remove the padding word at the end of the column
+        eq.18446744069414584320 if.true
+            sub.1
+        end
+
+        # OUTER: [c, words, addr, H, H, H, ...]
+        movdn.2 swap
+
+        # while(c>0)
+        sub.1 dup neq.0
+    end
+
+    # [H, H, H, ...]
+    drop drop drop
+end
+
+proc.test_is_zero_or_negative_one.0
+    push.0 exec.is_not_zero_or_negative_one not assert
+    push.0 sub.1 exec.is_not_zero_or_negative_one not assert
+    push.1 exec.is_not_zero_or_negative_one assert
+    push.2 exec.is_not_zero_or_negative_one assert
+    push.0 sub.2 exec.is_not_zero_or_negative_one assert
+end
```

Comment:
I think it would be good to put a separator above this like we do in the miden VM codebase.

Something like the following, with width to column 100
```
# TESTS 
# ==============================================================
```